src/backend/distributed/deparser/deparse_extension_stmts.c

```diff
@@ -86,54 +89,72 @@ DeparseCreateExtensionStmt(Node *node)
-	 * optionList as we will append "IF NOT EXISTS" clause regardless of
-	 * statement's content before propagating it to worker nodes.
-	 * We also do not care old_version for now.
-	 */
-	Value *schemaNameValue = GetExtensionOption(optionsList, "schema");
-
-	/* these can be NULL hence check before fetching the stored value */
-	Value *newVersionValue = GetExtensionOption(optionsList, "new_version");
-	Value *cascadeValue = GetExtensionOption(optionsList, "cascade");
+	if (createExtensionStmt->if_not_exists)
+	{
+		appendStringInfoString(buf, "IF NOT EXISTS ");
+	}
 
 	/*
-	 * We do not check for if schemaName is NULL as we append it in deparse
-	 * logic if it is not specified.
+	 * Up until here we have been ending the statement in a space, which makes it possible
+	 * to just append the quoted extname. From here onwards we will not have the string
+	 * ending in a space so appends should begin with a whitespace.
 	 */
-	const char *schemaName = strVal(schemaNameValue);
-	schemaName = quote_identifier(schemaName);
+	appendStringInfoString(buf, quote_identifier(createExtensionStmt->extname));
+	AppendCreateExtensionStmtOptions(buf, createExtensionStmt->options);
+	appendStringInfoString(buf, ";");
+}
 
-	appendStringInfo(buf, "CREATE EXTENSION IF NOT EXISTS %s WITH SCHEMA %s",
-					 extensionName, schemaName);
 
-	/* "new_version" may not be specified in CreateExtensionStmt */
-	if (newVersionValue)
+/*
+ * AppendCreateExtensionStmtOptions takes the option list of a CreateExtensionStmt and
+ * loops over the options to add them to the statement we are building.
+ *
+ * An error will be thrown if we run into an unsupported option, comparable to how
+ * postgres gives an error when parsing this list.
+ */
+static void
+AppendCreateExtensionStmtOptions(StringInfo buf, List *options)
+{
+	if (list_length(options) > 0)
 	{
-		const char *newVersion = strVal(newVersionValue);
-		newVersion = quote_identifier(newVersion);
-
-		appendStringInfo(buf, " VERSION %s", newVersion);
+		/* only append WITH if we actual will add options to the statement */
+		appendStringInfoString(buf, " WITH");
 	}
 
-	/* "cascade" may not be specified in CreateExtensionStmt */
-	if (cascadeValue)
+	/* Add the options to the statement */
+	DefElem *defElem = NULL;
+	foreach_ptr(defElem, options)
 	{
-		bool cascade = intVal(cascadeValue);
-
-		if (cascade)
+		if (strcmp(defElem->defname, "schema") == 0)
+		{
+			const char *schemaName = defGetString(defElem);
+			appendStringInfo(buf, " SCHEMA  %s", quote_identifier(schemaName));
+		}
+		else if (strcmp(defElem->defname, "new_version") == 0)
 		{
-			appendStringInfoString(buf, " CASCADE");
+			const char *newVersion = defGetString(defElem);
+			appendStringInfo(buf, " VERSION %s", quote_identifier(newVersion));
+		}
+		else if (strcmp(defElem->defname, "old_version") == 0)
+		{
+			const char *oldVersion = defGetString(defElem);
+			appendStringInfo(buf, " FROM %s", quote_identifier(oldVersion));
+		}
```

Comment:
Maybe we can add some tests for FROM (old_version) to propagate_extension.sql as well

Reply:
I have been thinking about this as well. This is causing a relative big portion of the change to not be tested.

The syntax itself is a bit ancient and I don't think we will ever use it :)

```
FROM old_version must be specified when, and only when, you are attempting to install an extension that replaces an “old style” module that is just a collection of objects not packaged into an extension
```

This was introduced when extensions became an actual thing. Before extensions you would just create some objects. To upgrade such modules to extensions an extension author could provide an automated script which could help you adopt such a module as an extension.

This syntax was introduced in 9.1, currently we do only support 11 and above, and I think old style modules are not used anymore.

Instead of adding a module that we can upgrade to an extension I think it will be easier to remove support for this, we could actually give an error when the syntax is used and hint people to remove the objects and perform a clean install of the extension.

What do you think?

Reply:
hmm, I see `dict_int` in contrib is a relatively small module with an `unpackaged` upgrade path. This might help us in adding coverage for this extension. Would require us to create 4 objects on the coordinator and workers before calling `FROM unpackaged`.

Reply:
I used `dict_int` as a test. It is a bit verbose as we need to create 4 objects on the coordinator and the worker for this to work, but it is good to test this.

Adding a node with an extension that is created like this works as well. It will not use the `FROM ...` syntax as it will create it from scratch as a clean install which is imho what should happen. The `FROM ...` syntax is only to upgrade manually created objects to an extension.

Reply:
Thanks for adding those tests, it's good to see that now we also support old style module updates as well